### PR TITLE
Improve error message when polygons contains  twice the same point in no-consecutive position

### LINKED ISSUE #40998: Self-intersecting shape reports strange error
In 6.6.2 I tried the following commands to index a shape with a hole that touches its edge:

```
PUT /i
{
  "mappings": {
    "_doc": {
      "properties": {
        "f": {
          "type": "geo_shape"
        }
      }
    }
  }
}

POST /i/_doc
{
  "f": {
    "coordinates": [
      [
        [
          0,
          0
        ],
        [
          2,
          0
        ],
        [
          1,
          1
        ],
        [
          1,
          2
        ],
        [
          2,
          0
        ],
        [
          2,
          3
        ],
        [
          0,
          3
        ],
        [
          0,
          0
        ]
      ]
    ],
    "type": "Polygon"
  }
}
```

This yielded the following response:

```
# 400 Bad Request
# {
#   "status": 400,
#   "error": {
#     "caused_by": {
#       "reason": "at least 4 polygon points required",
#       "type": "illegal_argument_exception"
#     },
#     "reason": "failed to parse field [f] of type [geo_shape]",
#     "root_cause": [
#       {
#         "reason": "failed to parse field [f] of type [geo_shape]",
#         "type": "mapper_parsing_exception"
#       }
#     ],
#     "type": "mapper_parsing_exception"
#   }
# }
```

I think this shape should indeed be rejected since it is self-intersecting at `[2,0]`, but the exception message seems wrong since there are enough points.

~I tried the same commands in 7.0.0-rc1 and this shape was accepted, which I also think shouldn't be the case.~ (edit, not true, see below)

The shape in question looks like this:

Screenshot 2019-04-09 at 09 38 22

Derived from the issue reported here: https://discuss.elastic.co/t/failed-to-create-valid-geo-shape/175975

### PR DESCRIPTION
**Pull request by iverase**
When a polygon contains a self-intersection due to have twice the same point in no-consecutive position it tries to split the polygon. During the split one of the polygons become invalid as it is not closed and an error is thrown which is not related to the real issue.

We capture now when this situation happen and throw a more meaningful error.

closes #40998

### CODE REVIEW

File: server/src/main/java/org/elasticsearch/common/geo/builders/PolygonBuilder.java [hunk 1]
```diff
@@ -479,10 +482,15 @@ private static int component(final Edge edge, final int id, final ArrayList<Edge
...
      */
-    private static Coordinate[] coordinates(Edge component, Coordinate[] coordinates) {
+    private static Coordinate[] coordinates(Edge component, Coordinate[] coordinates, double[] partitionPoint) {
         for (int i = 0; i < coordinates.length; i++) {
             coordinates[i] = (component = component.next).coordinate;
         }
+        // First and last coordinates must be equal
+        if (coordinates[0].equals(coordinates[coordinates.length - 1]) == false) {
+            throw new InvalidShapeException("Self-intersection at or near point ["
+                + partitionPoint[0] + "," + partitionPoint[1] + "," + partitionPoint[2] + "]");
```
Comment (imotov) [inline server/src/main/java/org/elasticsearch/common/geo/builders/PolygonBuilder.java]: I would add a check for `partitionPoint[2]` value and only print it if it is not `NULL_ORDINATE` (`NaN`).

**Review by imotov — APPROVED** [review]
I don't really like the style of passing arrays around and filling them in inside functions, but I cannot think of a better way to write it in less brittle way without major refactoring. Moreover this approach is already used through this class, so at least this is consistent. I think we should refactor this class at some point to make it easier to understand and a bit more functional. Meanwhile, LGTM with a nit-pick about always printing z-coordinate.

**Comment by iverase** [comment]
@imotov I do agree with you with the style but as you said, I did not find a better way without being too invasive. I hope we will be doing some refactoring at some point.

**Comment by iverase** [comment]
@elasticmachine retest this please

**Pull request merged**